File: Tree/HuffmanTree.cpp
#include<iostream>


/*

Push all the characters in ch[] mapped to corresponding frequncy freq[] in priority queue.
To create Huffman Tree, pop two nodes from priority queue.
Assign two popped node from priority queue as left and right child of new node.
Push the new node formed in priority queue.
Repeat all above steps untill size of priority queue becomes 1.
Traverse the only node in the priority queue for each character in ch[] and store the Huffman Code.
To print all the stored Huffman Code for every character in ch[].


*/

using namespace std;

typedef struct node {
	int data;
	node *left_child;
	node *right_child;
} node;


class queue {
	private:
		int capacity;
		node **arr;
		int rear;
		int front;
		int size;
	public:
		queue(int capacity) {
			this->capacity=capacity;
			this->arr=new node*[capacity];
			this->rear=-1;
			this->front=0;
			this->size=0;
		}

		int get_size() {
			return this->size;
		}

		void enqueue(node* x) {
			if(size==capacity) {
				cout << "Overflow!" << endl;
				return;
			}
			rear=(rear+1)%capacity;
			int index=size;
			if(size==0) {	
				index=rear;
			}
			for(int i=0;i<size;i++) {
				if(x->data<=arr[(front+i)%capacity]->data) {
					index=i;
					break;
				}
			}
			if(index!=size) {
				for(int i=size-1;i>=index;i--) {
					node* temp=arr[(front+i)%capacity];
					arr[(front+i+1)%capacity]=temp;
				}
			}
			arr[(front+index)%capacity]=x;
			size++;
		}

		node* dequeue() {
			if(size==0) {
				cout << "Underflow" << endl;
				exit(0);
			}
			node* temp=arr[front];
			size--;
			front=(front+1)%capacity;
			return temp;
		}

		void print() {
			cout << "\nCurrent Queue: " << endl;			
			for(int i=0;i<size;i++) {
				cout << arr[(front+i)%capacity]->data << endl;
			}
		}
		bool is_empty() {
			return this->size==0?true:false;
		}

};

void print_tree(node* root);

void print_huffman_tree(node* root, int path[], int size);

int main() {
	queue x(26);
	string s;
	cout << "Enter a string with lower case alphabets only: " <<endl;
	cin >> s;
	int freq[26];
	for(int i=0;i<26;i++) {
		freq[i]=0;
	}
	for(int i=0;i<s.length();i++) {
		freq[s[i]-'a']+=1;
	}	
	for(int i=0;i<s.length();i++) {
		if(freq[i]>0) {
			node* k=new node;
			k->data=freq[i];
			x.enqueue(k);
		}
	}
	x.print();
	int size=x.get_size();
	node *root;
	while(!x.is_empty()) {
		node* first=x.dequeue();
		node* second=x.dequeue();
	    node* new_node=new node;
		new_node->data=(first->data)+(second->data);
		new_node->right_child=first;
		new_node->left_child=second;
		if(!x.is_empty()) {
			x.enqueue(new_node);
		} else {
			root=new_node;
		}
	}
	int path[size];
	cout << "\n\nHuffman Tree: (Pre-Order Traversal)" <<endl;
	print_tree(root);
	cout << "\nHuffman Codes: " << endl;
	print_huffman_tree(root,path,0);

}


void print_tree(node* root) {
	if(root!=NULL) {
		cout << root->data << endl;
		print_tree(root->left_child);
		print_tree(root->right_child);
	}
}

void print_huffman_tree(node* root, int path[], int size) {
	if(root->left_child==NULL&&root->right_child==NULL) {
		cout << root->data << ": ";
		for(int i=0;i<size;i++) {
			cout << path[i]; 
		}
		cout << endl;
	}
	if(root->left_child!=NULL) {
			path[size]=0;
			size++;
			print_huffman_tree(root->left_child,path,size);
			size--;
	}
	if(root->right_child!=NULL) {
			path[size]=1;
			size++;
			print_huffman_tree(root->right_child,path,size);
			size--;
	}
	
}

File: src/main/java/in/sg/dsa/tree/HuffmanTree.java
package in.sg.dsa.tree;

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Scanner;

/**
 * Implements the Huffman coding algorithm for data compression.
 *
 * <p>Huffman coding is a lossless data compression algorithm that assigns variable-length codes to
 * input characters, lengths of the assigned codes are based on the frequencies of corresponding
 * characters. The most frequent character gets the smallest code and the least frequent character
 * gets the largest code.
 *
 * <p>Algorithm Steps: 1. Calculate the frequency of each character in the input string. 2. Create a
 * leaf node for each unique character and build a min-heap (PriorityQueue) of all leaf nodes. The
 * priority is determined by the frequency of the character. 3. Repeatedly extract the two nodes
 * with the minimum frequency from the heap. 4. Create a new internal node with a frequency equal to
 * the sum of the two nodes' frequencies. Make the first extracted node its left child and the
 * second extracted node its right child. 5. Add this new node back to the min-heap. 6. Repeat steps
 * 3-5 until the heap contains only one node, which is the root of the Huffman Tree. 7. Traverse the
 * tree from the root to generate the prefix codes for each character. (0 for left, 1 for right).
 *
 * <p>Time Complexity: O(n log n), where n is the number of unique characters. - Building the
 * frequency map: O(L) where L is the length of the input string. - Building the priority queue: O(n
 * log n) - Building the tree: O(n log n) as each extraction and insertion takes O(log n). -
 * Generating codes: O(n) to traverse the final tree.
 *
 * <p>Space Complexity: O(n) for the frequency map, priority queue, and the Huffman tree itself.
 */
public class HuffmanTree {

  private Node root;
  private final String text;
  private final Map<Character, String> huffmanCodes;

  // Inner node class
  private static class Node implements Comparable<Node> {
    char character;
    int frequency;
    Node left;
    Node right;

    Node(char character, int frequency, Node left, Node right) {
      this.character = character;
      this.frequency = frequency;
      this.left = left;
      this.right = right;
    }

    // A leaf node has a character, an internal node does not.
    boolean isLeaf() {
      return this.left == null && this.right == null;
    }

    @Override
    public int compareTo(Node other) {
      return this.frequency - other.frequency;
    }
  }

  public HuffmanTree(String text) {
    this.text = text;
    this.huffmanCodes = new HashMap<>();
    buildTree();
  }

  private void buildTree() {
    // 1. Calculate frequencies
    Map<Character, Integer> freqMap = new HashMap<>();
    for (char character : text.toCharArray()) {
      freqMap.put(character, freqMap.getOrDefault(character, 0) + 1);
    }

    // 2. Initialize the priority queue with leaf nodes
    PriorityQueue<Node> pq = new PriorityQueue<>();
    for (Map.Entry<Character, Integer> entry : freqMap.entrySet()) {
      pq.add(new Node(entry.getKey(), entry.getValue(), null, null));
    }

    // 3. Build the tree
    while (pq.size() > 1) {
      Node left = pq.poll();
      Node right = pq.poll();
      // Internal node has a dummy character and sum of frequencies
      Node parent = new Node('\0', left.frequency + right.frequency, left, right);
      pq.add(parent);
    }

    this.root = pq.poll();
    // 4. Generate codes
    generateCodes(root, "");
  }

  // 5. Recursive helper to generate codes
  private void generateCodes(Node node, String code) {
    if (node == null) {
      return;
    }
    // If it's a leaf node, it's one of our original characters
    if (node.isLeaf()) {
      huffmanCodes.put(node.character, code);
    }
    generateCodes(node.left, code + "0");
    generateCodes(node.right, code + "1");
  }

  public void printCodes() {
    System.out.println("\nHuffman Codes:");
    for (Map.Entry<Character, String> entry : huffmanCodes.entrySet()) {
      System.out.println("'" + entry.getKey() + "': " + entry.getValue());
    }
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    System.out.print("Enter the string to be compressed: ");
    String input = sc.nextLine();

    if (input == null || input.isEmpty()) {
      System.out.println("Input string is empty.");
      return;
    }

    HuffmanTree huffmanTree = new HuffmanTree(input);
    huffmanTree.printCodes();

    sc.close();
  }
}
